File: 프로그래머스/3/77486. 다단계 칫솔 판매/다단계 칫솔 판매.java
import java.util.*;

class Solution {
    public int[] solution(String[] enroll, String[] referral, String[] seller, int[] amount) {
        Map<String, Integer> total = new HashMap<>();
        Map<String, String> member = new HashMap<>();
        for(int i = 0; i < enroll.length; i++) 
            member.put(enroll[i], referral[i]);
        
        for(int i = 0; i < seller.length; i++) {
            String sName = seller[i];
            int sMoney = amount[i] * 100;
            
            while(sMoney > 0 && !sName.equals("-")) {
                total.put(sName, total.getOrDefault(sName, 0) + sMoney - (sMoney / 10));
                sName = member.get(sName);
                sMoney /= 10;
            }
        }
        
        
        
        int[] answer = new int[enroll.length];
        for(int i = 0; i < enroll.length; i++) {
            answer[i] = total.getOrDefault(enroll[i], 0);
        }
        return answer;
    }
}

File: 프로그래머스/3/77486. 다단계 칫솔 판매/다단계 칫솔 판매.py
def cal(name, price):
    global money, tree

    percent10 = int(price * 0.1)
    money[name] += price - percent10

    if percent10 == 0:
        return

    if name != '-':
        cal(tree[name], percent10)

def solution(enroll, referral, seller, amount):
    global money, tree

    money = dict()
    money['-'] = 0
    for name in enroll:
        money[name] = 0

    tree = dict()
    for child, parent in zip(enroll, referral):
        tree[child] = parent

    for name, price in zip(seller, amount):
        cal(name, price * 100)

    return [money[name] for name in enroll]